Fix Issue 14622 - documented unit test following undocumented symbol

ntrel opened the pull request:
Add diagnostic.

ntrel committed:
Fix Issue 14622 - documented unit test following undocumented symbol

Geod24 commented on compiler/src/dmd/parse.d:
```diff
@@ -491,6 +491,12 @@ class Parser(AST, Lexer = dmd.lexer.Lexer) : Lexer
                  */
                 if (mod.isRoot() && (global.params.useUnitTests || global.params.ddoc.doOutput || global.params.dihdr.doOutput))
                 {
+                    if (pAttrs.comment && *pLastDecl && !pLastDecl.comment)
+                    {
+                        error("Documented unittest found following undocumented symbol `%s`",
+                            pLastDecl.toPrettyChars());
+                        errorSupplemental(pLastDecl.loc, "`%s` declared here", pLastDecl.toChars());
+                    }
```
Let's not error out on documentation unless we are generating documentation.

Geod24 commented:
Looking at this from another PoV, why aren't public symbol put in the documentation regardless of their `///` ?
I use `///` everywhere because that IMO should be the default behavior.

ntrel committed:
Ignore if previous symbol is an undocumented unittest

There's no way to get the matching doc symbol from the previous
unittest.

ntrel committed:
Fix compilable example

ntrel committed:
Only error with -D

ntrel committed:
Fix detecting documented UT after undocumented UT; add rationale

Move check to doc.d.
Also remove unnecessary compile check that is already in `compilable`.

ntrel committed:
Found undocumented AssocArray with documented test!

ntrel commented:
@Geod24 For general symbols: sometimes you might write documentation for a symbol and then decide it's not ready, so you remove the doc comment. For nested doc comments inside that symbol it would be awkward to go through and undocument them, then when it's ready go through it again and remember to re-document all of them properly.

Documented unit tests are different because they are easier to accidentally break, I explained why in a comment in 89f79e7.

ntrel committed:
Move astTypeName unittest under doc symbol

ntrel committed:
Don't call writeFile if there were errors

ntrel committed:
Fix CoreDoc -> CoreDdoc

ntrel committed:
Move global.errors check after last error

ntrel committed:
Fix Thread UT

ntrel commented on druntime/src/core/thread/osthread.d:
```diff
@@ -186,6 +186,16 @@ else
+    return cast(Thread) cast(void*) t;
+}
+
+private extern(D) static void thread_yield() @nogc nothrow
```
This is private and not used anywhere in this module. Presumably I should remove it?

ntrel commented:
The buildkite `ae` failure seems unrelated:
> std.socket.SocketOSException@std/socket.d(2826): Unable to bind socket: Cannot assign requested address
 

RazvanN7 commented:
`ae` doesn't fail anymore, however, we have the documentation tester failing:

```
/home/dtest/DAutoTest/work/tmp/ddoc_preprocessor_64396295aa5d4f108caabf7e23507f68/std/file.d(1985): Error: Documented unittest found following undocumented symbol `existsImpl` /home/dtest/DAutoTest/work/tmp/ddoc_preprocessor_64396295aa5d4f108caabf7e23507f68/std/file.d(1947):        `existsImpl` declared here
```

which is valid.
